### PR TITLE
Add more WebAssembly bindings

### PR DESCRIPTION
**Pull request by RazerM**
_Should_ check off the following:

- #275
  - [ ] `WebAssembly`
    - [x] `WebAssembly.compile()`
  - [x] `WebAssembly.Module`
    - [x] `WebAssembly.Module.customSections()`
    - [x] `WebAssembly.Module.exports()`
    - [x] `WebAssembly.Module.imports()`
  - [ ] `WebAssembly.Table`
    - [x] `WebAssembly.Table.prototype.length`
  - [x] `WebAssembly.CompileError`
  - [x] `WebAssembly.LinkError`
  - [x] `WebAssembly.RuntimeError`
- #670 
  - [x] `WebAssembly.Module`
  - [x] `WebAssembly.Table`

Please point out anything that doesn't look quite right.

### CODE REVIEW

File: crates/js-sys/Cargo.toml [hunk 1]
```diff
@@ -18,7 +18,9 @@ test = false
 doctest = false
 
 [dependencies]
+futures = "0.1.20"
```
[the remark this answers is not in the document]
Comment (RazerM) [reply crates/js-sys/Cargo.toml]: oops, yes

File: crates/js-sys/src/lib.rs [hunk 2]
```diff
@@ -2843,19 +2843,163 @@ extern "C" {
...
+        #[derive(Clone, Debug)]
+        pub type Module;
+
+        /// A `WebAssembly.Module` object contains stateless WebAssembly code
+        /// that has already been compiled by the browser and can be
+        /// efficiently shared with Workers, and instantiated multiple times.
+        ///
+        /// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WebAssembly/Module
+        #[wasm_bindgen(constructor, js_namespace = WebAssembly)]
+        pub fn new(buffer_source: &JsValue) -> Module;
```
[the remark this answers is not in the document]
Comment (RazerM) [reply crates/js-sys/src/lib.rs]: I had already tried to add `catch` for the constructors that can fail, but it seems it isn't supported:

```
error: paths with type parameters are not supported yet
    --> crates/js-sys/src/lib.rs:2955:48
     |
2955 |         pub fn new(buffer_source: &JsValue) -> Result<Module, JsValue>;
     |                                                ^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to previous error
```

File: crates/js-sys/src/lib.rs [hunk 3]
```diff
@@ -2843,19 +2843,163 @@ extern "C" {
...
+        #[wasm_bindgen(js_namespace = WebAssembly, extends = Object)]
+        #[derive(Clone, Debug)]
+        pub type Table;
+
+        /// The `WebAssembly.Table()` constructor creates a new `Table` object
+        /// of the given size and element type.
+        ///
+        /// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WebAssembly/Table
+        #[wasm_bindgen(constructor, js_namespace = WebAssembly)]
+        pub fn new(table_descriptor: &JsValue) -> Table;
```
[the remark this answers is not in the document]
Comment (RazerM) [reply crates/js-sys/src/lib.rs]: Yes, I'll change to `&Object`

File: crates/js-sys/src/lib.rs [hunk 4]
```diff
@@ -2848,19 +2848,163 @@ extern "C" {
...
+        /// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WebAssembly/Module/customSections
+        #[wasm_bindgen(static_method_of = Module, js_namespace = WebAssembly, js_name = customSections, catch)]
+        pub fn custom_sections(module: &Module, sectionName: &str) -> Result<Array, JsValue>;
+
+        /// The `WebAssembly.exports()` function returns an array containing
+        /// descriptions of all the declared exports of the given `Module`.
+        ///
+        /// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WebAssembly/Module/exports
+        #[wasm_bindgen(static_method_of = Module, js_namespace = WebAssembly, catch)]
+        pub fn exports(module: &Module) -> Result<Array, JsValue>;
```
Comment (alexcrichton) [inline crates/js-sys/src/lib.rs]: Do this, `imports`, and `custom_sections` need `catch`? (or do they regularly return errors?)

File: crates/js-sys/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (alexcrichton) [reply crates/js-sys/src/lib.rs]: Hm true! Mind commenting this out in the meantime while we fix that? Also, mind filing an issue for that? (I think it's specifically related to constructors and `catch`)

[the remark this answers is not in the document]
Comment (alexcrichton) [reply crates/js-sys/src/lib.rs]: Yeah just as a stopgap until we implement it! (which with an issue shouldn't take too long to track down!)

File: crates/js-sys/src/lib.rs (hunk 4 above)
Comment (RazerM) [reply crates/js-sys/src/lib.rs]: Right, `catch` can be removed because rust prevents the `TypeError`

**Comment by alexcrichton** [comment]
I've got https://github.com/rustwasm/wasm-bindgen/pull/740 to fix the problem of combining `catch` and `constructor`

**Comment by fitzgen** [comment]
> I've got #740 to fix the problem of combining catch and constructor

This is fixed and merged into master now, so we should be able to enable those constructors now too.

**Comment by alexcrichton** [comment]
Awesome, thanks @RazerM!

**Pull request merged**